resource/kind.go

```diff
@@ -0,0 +1,431 @@
+	GVK schema.GroupVersionKind
+	// PluralKind is the plural name of the Kind, returned via Plural()
+	PluralKind string
+	// KindScope is the scope of the Kind, returned by Scope()
+	KindScope SchemaScope
+	// ZeroObject is the *UntypedObject returned by ZeroValue(). This only needs to be non-nil if you do not want to use
+	// an empty *UntypedObject as the ZeroValue() return.
+	ZeroObject *UntypedObject
+}
+
+func (u *UntypedKind) Group() string {
+	return u.GVK.Group
+}
+func (u *UntypedKind) Version() string {
+	return u.GVK.Version
+}
+func (u *UntypedKind) Kind() string {
+	return u.GVK.Kind
+}
+func (u *UntypedKind) Plural() string {
+	if u.PluralKind == "" {
+		return strings.ToLower(u.GVK.Kind) + "s"
+	}
+	return u.PluralKind
+}
+func (u *UntypedKind) ZeroValue() Object {
+	if u.ZeroObject != nil {
+		return u.ZeroObject.Copy()
+	}
+	return &UntypedObject{}
+}
+func (u *UntypedKind) Scope() SchemaScope {
+	if u.KindScope == SchemaScope("") {
+		return NamespacedScope
+	}
+	return u.KindScope
+}
+
+// Read reads in kubernetes JSON bytes for any kind and returns an *UntypedKind for those bytes
+func (u *UntypedKind) Read(in io.Reader) (Object, error) {
+	obj := &UntypedObject{}
+	if err := json.NewDecoder(in).Decode(&obj); err != nil {
+		return nil, err
+	}
+	return obj, nil
+}
+
+// ReadInto reads in kubernetes JSON bytes for any kind and attempts to unmarshal them into the provided `into` Object2
+func (u *UntypedKind) ReadInto(in io.Reader, into Object) error {
+	if in == nil {
+		return fmt.Errorf("in io.Reader cannot be nil")
+	}
+	if into == nil {
+		return fmt.Errorf("into Object cannot be nil")
+	}
+	// TODO: make this a better unmarshal?
+	return json.NewDecoder(in).Decode(&into)
+}
+
+// Write takes in any Object2 and outputs a kubernetes object JSON
+func (u *UntypedKind) Write(obj Object, out io.Writer) error {
+	m := make(map[string]any)
+	m["apiVersion"], m["kind"] = obj.GetObjectKind().GroupVersionKind().ToAPIVersionAndKind()
+	m["metadata"] = metav1.ObjectMeta{
+		Name:                       obj.GetName(),
+		GenerateName:               obj.GetGenerateName(),
+		Namespace:                  obj.GetNamespace(),
+		SelfLink:                   obj.GetSelfLink(),
+		UID:                        obj.GetUID(),
+		ResourceVersion:            obj.GetResourceVersion(),
+		Generation:                 obj.GetGeneration(),
+		CreationTimestamp:          obj.GetCreationTimestamp(),
+		DeletionTimestamp:          obj.GetDeletionTimestamp(),
+		DeletionGracePeriodSeconds: obj.GetDeletionGracePeriodSeconds(),
+		Labels:                     obj.GetLabels(),
+		Annotations:                obj.GetAnnotations(),
+		OwnerReferences:            obj.GetOwnerReferences(),
+		Finalizers:                 obj.GetFinalizers(),
+		ManagedFields:              obj.GetManagedFields(),
+	}
+	m["spec"] = obj.GetSpec()
+	for k, v := range obj.GetSubresources() {
+		m[k] = v
+	}
+	return json.NewEncoder(out).Encode(m)
+}
```

Comment:
I think we should separate data types (e.g. `UntypedKind` / `TypedKind`) from serialisation logic. The latter should be implemented separately.

Reply:
So right now the separation is that a `Schema` is information about the data type, and a `Kind` combines that with serialization logic. Perhaps we have both `UntypedSchema`/`TypedSchema` in addition to the `UntypedKind`/`TypedKind`? The combination with serialization logic is mainly so that serialization isn't incumbent upon clients (like the `NegotiatedSerializer` in the k8s client), given that kind implementations may vary in logic (the easiest example right now is legacy thema support). 

Perhaps we also explicitly create typed/untyped `KindReadWriter` implementations, which explicitly say that they use bare-bones marshal/unmarshal, and then call out that `TypedKind`/`UntypedKind` use them?

Reply:
We could also change nomenclature so that `Schema` -> `Kind`, and `Kind` -> `SerializableKind`, or something to that effect, to make things clearer.

Reply:
> The combination with serialization logic is mainly so that serialization isn't incumbent upon clients (like the NegotiatedSerializer in the k8s client), given that kind implementations may vary in logic (the easiest example right now is legacy thema support).

I'm not sure I follow the line of thought here. Why is forcing codecs on the clients a problem? At the end of the day only the clients have enough information to be able to (de)serialise from / to specific wire formats, no?

Reply:
I keep going in circles with this in my head. We've gone from (de)serialization being controlled by the `Object` implementation itself, to now externalizing (de)serialization into the container for schema information so it can be better re-used and customized, I keep thinking it might be nice to further externalize it, but we do need to keep the option open for a particular kind (GVK) to (de)serialize differently, I _think_. But I'm also not adverse to what is essentially a `NegotiatedSerializer`-type implementation where _all_ kinds for a client (de)serialize the same way, it would just add a new restriction to the SDK that hasn't existed prior (and kind of throws a wrench in thema support, but I might be able to work around it).

This is _also_ something we could, in theory, change to later, by removing the (de)serialization logic from `Kind`, though it would be a breaking change, it would be no more a breaking change than doing the same thing right now, so we could also punt this conversation to a future time.

Reply:
I'm OK with keeping the codec logic in the Kind. I suggest removing all other fields from structs though and instead delegating them to `Schema`, i.e. something like:

```golang
type Codec interface {
  Encode(io.Writer, Object) error
  Decode(io.Reader, Object) error
}

type Kind struct {
  Schema Schema
  Codecs map[KindEncoding]Codec
}
```

Reply:
That looks like a good approach, thank you! I think this suggested approach is much clearer than the mess I've gotten myself into.